### PR TITLE
JBPM-9238 Stunner - Allow the introduction of emails in Tasks notifications

### PR DESCRIPTION
**Pull request by hasys**
Hi @romartin, @elguardian, @afalhambra, @LuboTerifaj, @gmunozfe,

there is Stunner part of the introduction of emails in Tasks notifications. It is still missing validation (see below Validation part) but it is ready for End to End tests already. Thank you!

### Related PRs are here
* droolsjbpm-knowledge: https://github.com/kiegroup/droolsjbpm-knowledge/pull/445
* JBPM: https://github.com/kiegroup/jbpm/pull/1684
image

## How Validation is done
* Split emails string by `,` 
* execute [**OWASP Validation Regex**](https://owasp.org/www-community/OWASP_Validation_Regex_Repository) `^[a-zA-Z0-9_+&*-]+(?:\.[a-zA-Z0-9_+&*-]+)*@(?:[a-zA-Z0-9-]+\.)+[a-zA-Z]{2,7}$` on each part.
* If one of the part is not correct according to the validation, show a message on OK button click:
image

### WAR file for testing
https://drive.google.com/file/d/1xvT7s00RipwKlb8Y25-ZX1rXaKYuIYCM/view?usp=sharing

### CODE REVIEW

**Comment by hasys** [comment]
Jenkins execute full downstream build

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html [hunk 1]
```diff
@@ -89,11 +89,11 @@
         <p style="font-weight: bold;">Message</p>
         <div class="form-group">
             <div class="col-sm-3" style="margin-left: -20px;">
-                <p>From</p>
+                <p style="margin: 0">From</p>
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html:92]: maybe we can fix also the i18n in this PR

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html [hunk 2]
```diff
@@ -89,11 +89,11 @@
...
                 <select data-field="liveSearchFromDropDown" id="from" style="width: 200px;"></select>
             </div>
             <div class="col-sm-3">
-                <p>Reply to (optional)
+                <p style="margin: 0">Reply to (optional)
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html:96]: maybe we can fix also the i18n in this PR

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html [hunk 3]
```diff
@@ -105,15 +105,19 @@
     <div class="row">
         <div class="form-group">
             <div class="col-sm-3" style="margin-left: -20px;">
-                <p>To: user(s)</p>
+                <p style="margin: 10px 0 0 0">To: user(s)</p>
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html:108]: maybe we can fix also the i18n in this PR

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html [hunk 4]
```diff
@@ -105,15 +105,19 @@
...
                 <select data-field="multipleSelectorInputUsers" id="toUsers" style="width: 200px;"></select>
             </div>
             <div class="col-sm-3">
-                <p>To: group(s)</p>
+                <p style="margin: 10px 0 0 0">To: group(s)</p>
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html:112]: maybe we can fix also the i18n in this PR

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html [hunk 5]
```diff
@@ -105,15 +105,19 @@
...
                 <select data-field="multipleSelectorInputGroups" id="toGroups" style="width: 200px;"></select>
             </div>
         </div>
     </div>
+    <div class="row">
+        <p style="margin: 10px 0 0 0;">To: email(s)</p>
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html:118]: maybe we can fix also the i18n in this PR

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html [hunk 6]
```diff
@@ -105,15 +105,19 @@
...
+        <input data-field="emails" type="text" class="form-control" max='99' placeholder='Enter emails separated by comma.' style='width:500px;'>
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html:119]: maybe we can fix also the i18n in this PR

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html [hunk 7]
```diff
@@ -105,15 +105,19 @@
...
+    </div>
     <div class="row">
         <input data-field="subject" type="text" class="form-control" max='99' placeholder='Enter the subject here.' style=' width:500px;    margin-top: 20px;'>
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html:122]: maybe we can fix also the i18n in this PR

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.java [hunk 8]
```diff
@@ -426,10 +437,10 @@ public void createOrEdit(NotificationWidgetView parent, NotificationRow row) {
         }
 
         if (row.getType() != null) {
-            if (row.getType().equals(NotificationType.NotCompletedNotify)) {
+            if (row.getType().equals(NOT_COMPLETED_NOTIFY)) {
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.java:440]: by flipping the condition as NOT_COMPLETED_NOTIFY.equals(row.getType()) there is no need for the NPE check above

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.java [hunk 9]
```diff
@@ -426,10 +437,10 @@ public void createOrEdit(NotificationWidgetView parent, NotificationRow row) {
...
                 notCompletedInput.checked = true;
                 notStartedInput.checked = false;
-            } else if (row.getType().equals(NotificationType.NotStartedNotify)) {
+            } else if (row.getType().equals(NOT_STARTED_NOTIFY)) {
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.java:443]: same here, by flipping the condition no need for the NPE check

**Comment by romartin** [comment]
@hasys no worries for the builds on this PR yet, as it cannot yet be tested as it depends on some other PR for being merged, right? As commented, let's just wait for the dependant PRs to be merged and then move this to a normal PR, and let's start the review again. 
Does it makes sense? Plz let us know when this happens.
Thanks!!

**Comment by hasys** [comment]
Hi @romartin,

yes, I think I finished here and waiting for the next review :) 

What I did after your last review:
* All layout issues I found in Notification Editor are fixed
* All i18n issues in Notification Editor HTML file are fixed, it is all localized now (just missing translations)
* Fixed coverage issues provided by SonarCloud (I did some refactoring on View/Presenter as well since SonarCloud not calculating GwtMocked coverage correctly I made presenter as normal instance and moved logic (I touched) to the presenter, so in View now mostly logic I didn't touch or declarative only methods).

@afalhambra, @LuboTerifaj I didn't provide Full Downstream Build since there is still missing part in Business Central so it is not possible to perform End to End testing until all other PRs will be merged.

Does it make sense? Should we test UI part only when other PRs will be merged? If not, how will we build Business Central? Since Jenkins can't build Full Downstream Build using PRs from different people.

**Comment by hasys** [comment]
Jenkins retest please

**Comment by hasys** [comment]
Hi @romartin, @LuboTerifaj  I added more tests, and now SonarCloud complains about duplication only, but duplication caused by duplicated Marshallers, so not much I can do here.

See overall status in [previous comment](https://github.com/kiegroup/kie-wb-common/pull/3358#issuecomment-663110955), thank you!

**Comment by hasys** [comment]
Hi @romartin, @afalhambra,

all related PRs are merged and I rebased to master and built a WAR for End To End tests, this PR is ready for review. Thank  you! 

For WAR file see at the bottom https://github.com/kiegroup/kie-wb-common/pull/3358#issue-451032496

**Comment by hasys** [comment]
Jenkins retest please

**Comment by hasys** [comment]
Jenkins execute compile downstream build

**Comment by hasys** [comment]
Jenkins execute compile downstream build

**Comment by romartin** [comment]
Hey @hasys 

I'd been looking again at code updates, looks great, great job! :+1: 

Just a couple of Qs before finally approving please:

1. Maybe makes sense making the text area used for the body resizable (as an example look at the text area for any name or documentation field)
2. For the email realted fields to users and groups, I understand we're not querying the user management services right? (which is present for BC, and "mocked" to empty value results in Kogito)

Anyway for me this PR is ok to go, but would like to clarify both improvements above to see if makes sense to also include.

Thanks!

**Comment by hasys** [comment]
Hi @romartin good questions, thank you!
1. It is a good possible improvements, but since we have a green build I prefer to report it as separate issue since it is not related to this jira issue. If we will have more fixes to do in this PR than we can do it together since change is trivial. But let's concentrate on merge now if no more critical issues found and improve in separate PR to be safe for the release.
2. No, I didn't use any services. If I correct Users and Groups are provided by EAP/Wildfly user/groups but it doesn't provide any emails. So emails field is just a textbox field without any addition services. Let me know if I am wrong and there is a service for emails. Also I think we can add call to the service as separate PR as well.

Thank you for good questions @romartin!

**Comment by romartin** [comment]
Jenkins execute full downtream build

**Comment by romartin** [comment]
@hasys lemme know when it's ready to merge!!! :+1:  FDB is running

**Comment by hasys** [comment]
Hi @romartin, this PR has green Compile Downstream Build so it is ready for merge. Thank you!

**Pull request merged**